### PR TITLE
Add api search

### PR DESCRIPTION
**Pull request by kand617**
Added Search API using Hibernate Search

New endpoints
/apis - > Returns All Apis from the database (excluding swagger descriptions)
/apis?query="hello" -> Search apis based on query string. 

Other Notes:
All Deprecated Apis are filtered from search.

### CODE REVIEW

**Comment by sudhirtumati** [comment]
There are unit test failures

**Comment by sudhirtumati** [comment]
Fixed unit test failures with commit d3d8500

File: .gitignore [hunk 1]
```diff
@@ -165,4 +165,5 @@ buildNumber.properties
 !/.mvn/wrapper/maven-wrapper.jar
 
 
-# End of https://www.gitignore.io/api/java,maven,eclipse,intellij
\ No newline at end of file
+# End of https://www.gitignore.io/api/java,maven,eclipse,intellij
+io\.apicollab\.server\.domain\.Api/
```
[the remark this answers is not in the document]
Comment (kand617) [reply .gitignore]: awesome

File: src/main/java/io/apicollab/server/config/HibernateSearchConfig.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
+package io.apicollab.server.config;
```
[the remark this answers is not in the document]
Comment (kand617) [reply src/main/java/io/apicollab/server/config/HibernateSearchConfig.java]: web.commons? Not sure which package would be the best

File: src/main/java/io/apicollab/server/config/HibernateSearchConfig.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+     */
+    @Override
+    @Transactional
+    public void onApplicationEvent(final ApplicationReadyEvent event) {
+        try {
+            FullTextEntityManager fullTextEntityManager =
+                    Search.getFullTextEntityManager(entityManager);
+            fullTextEntityManager.createIndexer().startAndWait();
+        } catch (InterruptedException e) {
+            System.out.println(
```
[the remark this answers is not in the document]
Comment (kand617) [reply src/main/java/io/apicollab/server/config/HibernateSearchConfig.java]: done

File: src/main/java/io/apicollab/server/repository/ApiRepository.java [hunk 4]
```diff
@@ -13,4 +17,7 @@
     Optional<Api> findByApplicationIdAndNameAndVersion(String applicationId, String name, String version);
 
     Collection<Api> findByApplicationId(String applicationId);
+
+    @Query("FROM Api WHERE status in (:statusCodes)")
+    Collection<Api> getAllByStatus(@Param("statusCodes") List<ApiStatus> statusCodes);
```
[the remark this answers is not in the document]
Comment (kand617) [reply src/main/java/io/apicollab/server/repository/ApiRepository.java]: Ah didnt realise the "In" suffix existed.  

Done

File: src/main/java/io/apicollab/server/config/HibernateSearchConfig.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (sudhirtumati) [reply src/main/java/io/apicollab/server/config/HibernateSearchConfig.java]: To be honest, even I am not quite sure. Let's leave it in this package for now.

File: src/main/java/io/apicollab/server/controller/ApiController.java [hunk 5]
```diff
@@ -56,6 +57,9 @@ public ApiListDTO getApplicationApis(@PathVariable String applicationId) {
 
     @GetMapping("/apis")
     public ApiListDTO getApis(@RequestParam(name = "query", required = false) String query) {
+        if (StringUtils.isBlank(query)) {
```
Comment (sudhirtumati) [inline src/main/java/io/apicollab/server/controller/ApiController.java]: Wouldn't it be better to have a dedicated `/search` endpoint with required parameter `query`? At least it helps us to get rid off this **_evil_** if :)

Comment (kand617) [reply src/main/java/io/apicollab/server/controller/ApiController.java]: Other than the *evil* if, is there a more reason? I am not a fan of /search endpoints, unless we really need POST: /search....

Comment (sudhirtumati) [reply src/main/java/io/apicollab/server/controller/ApiController.java]: Not really. I would prefer get with filters and a search to be separate endpoints. Actually, I like the idea `POST /search` :)

Anyhow the current search is very limited in functionality. May be we can create a separate issue to improve search functionality to be able to filter results further (by status, application, version, ...) and make improvements as part of that issue.

File: src/main/java/io/apicollab/server/controller/ApiController.java [hunk 6]
```diff
@@ -56,6 +55,15 @@ public ApiListDTO getApplicationApis(@PathVariable String applicationId) {
         return ApiListDTO.builder().apis(apiMapper.toDtos(apiService.findByApplication(applicationId))).build();
     }
 
+    @GetMapping("/apis")
+    public ApiListDTO getApis(@RequestParam(name = "query", required = false) String query) {
```
Comment (koushiknama) [inline src/main/java/io/apicollab/server/controller/ApiController.java]: please sanitize the ```query```by validating against accepted/white listed characters.
To start with, avoid special characters. use regex ```[a-zA-Z0-9\-#\.\(\)\/%&\s]{0,19}```

File: src/test/java/io/apicollab/server/service/ApiSearchTests.java [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+import org.springframework.test.context.junit4.SpringRunner;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+@RunWith(SpringRunner.class)
+@SpringBootTest
+public class ApiSearchTests {
```
Comment (koushiknama) [inline src/test/java/io/apicollab/server/service/ApiSearchTests.java]: Only positive test scenarios are addressed, please list out negative scenarios and make sure the functionality doesn't break during those.

File: src/main/java/io/apicollab/server/controller/ApiController.java (hunk 5 above)
Comment (kand617) [reply src/main/java/io/apicollab/server/controller/ApiController.java]: OK Lets create /search?q="phrase" and later we add  extra query param property filters

**Pull request merged**